lightning/backend/local.go

```diff
@@ -1162,24 +1272,50 @@ func (local *local) ImportEngine(ctx context.Context, engineUUID uuid.UUID) erro
 	return nil
 }
 
-func (local *local) CleanupEngine(ctx context.Context, engineUUID uuid.UUID) error {
+func (local *local) ResetEngine(ctx context.Context, engineUUID uuid.UUID) error {
+	// the only way to reset the engine + reclaim the space is to delete and reopen it 🤷
+	exists, err := local.cleanupEngineIfExists(ctx, engineUUID)
+	if err != nil || !exists {
+		return err
+	}
+	return local.OpenEngine(ctx, engineUUID)
+}
+
+// cleanupEngineIfExists returns if the engine with given UUID exists. If the
+// engine already exists, it will be deleted.
+func (local *local) cleanupEngineIfExists(ctx context.Context, engineUUID uuid.UUID) (exists bool, err error) {
 	// release this engine after import success
 	engineFile, ok := local.engines.Load(engineUUID)
-	if ok {
-		localEngine := engineFile.(*LocalFile)
-		err := localEngine.Close()
-		if err != nil {
-			return err
-		}
-		err = localEngine.Cleanup(local.localStoreDir)
-		if err != nil {
-			return err
-		}
-		local.engines.Delete(engineUUID)
-	} else {
+	if !ok {
 		log.L().Warn("could not find engine in cleanupEngine", zap.Stringer("uuid", engineUUID))
+		return false, nil
 	}
-	return nil
+
+	localEngine := engineFile.(*LocalFile)
+	localEngine.lock(importMutexStateClose)
+	defer localEngine.unlock()
+
+	// since closing the engine causes all subsequent operations on it panic,
+	// we make sure to delete it from the engine map before calling Close().
+	// (note that Close() returning error does _not_ mean the pebble DB
+	// remains open/usable.)
+	local.engines.Delete(engineUUID)
```

Comment:
seems `pebble.DB` of engines are flushed when `CloseEngine`, while `CloseEngine` didn't check the lock of `LocalFile` to prevent `pebble.DB.Close` in `cleanupEngineIfExists`

Reply:
Agree, seems `LocalFile.Close` should also acquire the mutex first